src/rust/catmem/mod.rs

```diff
@@ -121,14 +141,12 @@ impl CatmemLibOS {
     /// Closes a memory queue.
     pub fn close(&mut self, qd: QDesc) -> Result<(), Fail> {
         trace!("close() qd={:?}", qd);
-        match self.rings.remove(&qd) {
-            Some(pipe) => {
-                self.push_eof(pipe.buffer())?;
-                self.qtable.free(qd);
-                Ok(())
-            },
-            None => Err(Fail::new(libc::EBADF, "invalid queue descriptor")),
-        }
+        match self.qtable.get(&qd) {
+            Some(queue) => self.push_eof(queue.get_pipe().buffer())?,
+            None => return Err(Fail::new(libc::EBADF, "invalid queue descriptor")),
```

Comment:
Yes, previously we always removed from the queue table, so we didn't have the possibility that it would fail. Now we check if the queue exists. Then try to push an EOF and if that succeeds, we free the queue descriptor.